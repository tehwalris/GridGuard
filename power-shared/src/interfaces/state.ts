export type DateNumber = number; // milliseconds, to pass to new Date(...)

export interface State {
  users: User[];
  toggles: MachineToggle[];
  simulation: SimulationState;
}

export interface User {
  id: string;
  player?: number | null;
}

export interface MachineToggle {
  key: string;
  powered: boolean;
}

export interface SimulationState {
  recentPowerConsumption: number[];
}
