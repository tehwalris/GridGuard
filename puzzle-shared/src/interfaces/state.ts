export type DateNumber = number; // milliseconds, to pass to new Date(...)

export interface State {
  phaseState: PhaseState;
  gameSettings: GameSettings;
  puzzleDefinition: PuzzleDefinition;
  solution: Solution;
  players: Player[]; // 4 elements
  users: User[];
}

export enum Phase {
  PreGame = "PreGame",
  Distributing = "Distributing",
  Solving = "Solving",
  PostGame = "PostGame",
}

export type PhaseState =
  | PreGameState
  | DistributingState
  | SolvingState
  | PostGameState;

export interface PreGameState {
  phase: Phase.PreGame;
  showImagePreviewNow: boolean;
}

export interface DistributingState {
  phase: Phase.Distributing;
  pieceStartTime: DateNumber; // reset after each piece is distributed
  distributor: number; // index of player
  toDistribute: number[][]; // toDistribute[playerIndex] is an array of piece indices, pop() to get next piece
  totalToDistribute: number; // sum of all toDistribute length before any distribution occurs
}

export interface SolvingState {
  phase: Phase.Solving;
  startTime: DateNumber;
  numberOfSwaps: number;
}

export interface PostGameState {
  phase: Phase.PostGame;
  numberOfSwaps: number;
  totalGameTime: number; //in seconds
}

export interface User {
  id: string;
  player?: number | null;
}

export interface PuzzleDefinition {
  pieces: PieceDefinition[]; // row major
  size: GridSize; // in pieces
  imageURL: string;
  tileHeight: number; //pixels
  tileWidth: number; //pixels
}

export interface PieceDefinition {}

export type GridPlacement = (number | null)[][]; // value is index into PuzzleDefinition.pieces

export interface Solution {
  pieces: GridPlacement;
}

export interface Player {
  name: string;
  tempSpace: GridPlacement;
  recentReceives: RecentReceive[];
  finished: boolean;
}

export interface RecentReceive {
  piece: number;
  fromPlayerName: string;
}

export interface GridSize {
  height: number;
  width: number;
}

export interface GridLocation {
  row: number;
  column: number;
}

export interface GridLocationRange {
  from: GridLocation; // inclusive
  to: GridLocation; // inclusive; must be >= from
}

export enum NumberOfTiles {
  Few = "Few",
  Medium = "Medium",
  Many = "Many",
}

export enum DistributionTime {
  Short = "Short",
  Medium = "Medium",
  Long = "Long",
}

export interface GameSettings {
  image: ImageWithSize;
  roughNumberOfTiles: NumberOfTiles;
  distributionTime: DistributionTime;
  enableImagePreview: boolean;
}

export interface ImageWithSize {
  url: string;
  width: number;
  height: number;
}
