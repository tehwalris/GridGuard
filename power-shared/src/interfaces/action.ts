import { Device, GridPowerConsumption } from "./state";

export type Action =
  | AddUserAction
  | RemoveUserAction
  | SetToggleAction
  | StartEventAction
  | EndEventAction
  | TickSimulationAction;

export enum ActionType {
  AddUser = "AddUser",
  RemoveUser = "RemoveUser",
  SetToggle = "SetToggle",
  StartEvent = "StartEvent",
  EndEvent = "EndEvent",
  TickSimulation = "TickSimulation",
}

export interface AddUserAction {
  type: ActionType.AddUser;
  userId: string;
}

export interface RemoveUserAction {
  type: ActionType.RemoveUser;
  userId: string;
}

export interface SetToggleAction {
  type: ActionType.SetToggle;
  key: string;
  powered: boolean;
}
export interface StartEventAction {
  type: ActionType.StartEvent;
}
export interface EndEventAction {
  type: ActionType.EndEvent;
}
export interface TickSimulationAction {
  type: ActionType.TickSimulation;
  devices: Device[]; // only a subset of devices for perf reasons
  powerConsumption: GridPowerConsumption; // from all devices, not just the subset
}
