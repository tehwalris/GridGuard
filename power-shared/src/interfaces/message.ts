import { State } from "..";
import { Action } from "./action";

export interface LogEntry {
  id: number;
  undoKey?: string;
  action: Action;
}

export enum MessageType {
  LogEntryServer = "LogEntryServer",
  RemapEntryServer = "RemapEntryServer",
  RejectEntryServer = "RejectEntryServer",
  InitialServer = "InitialServer",
  ReportUndoServer = "ReportUndoServer",
  SubmitEntryClient = "SubmitEntryClient",
  RequestUndoClient = "RequestUndoClient",
}

export type ServerMessage =
  | InitialServerMessage
  | LogEntryServerMessage
  | RemapEntryServerMessage
  | RejectEntryServerMessage
  | ReportUndoServerMessage;

export interface InitialServerMessage {
  type: MessageType.InitialServer;
  initialState: State;
  userId: string;
}

export interface LogEntryServerMessage {
  type: MessageType.LogEntryServer;
  entry: LogEntry;
}

export interface RemapEntryServerMessage {
  type: MessageType.RemapEntryServer;
  oldId: number;
  entry: LogEntry;
}

export interface RejectEntryServerMessage {
  type: MessageType.RejectEntryServer;
  entryId: number;
  error: string;
}

export interface ReportUndoServerMessage {
  type: MessageType.ReportUndoServer;
  undoKey: string;
  finalEntryId: number;
  finalState: State;
}

export type ClientMessage = SubmitEntryClientMessage | RequestUndoClientMessage;

export interface SubmitEntryClientMessage {
  type: MessageType.SubmitEntryClient;
  entry: LogEntry;
}

export interface RequestUndoClientMessage {
  type: MessageType.RequestUndoClient;
  undoKey: string;
}
