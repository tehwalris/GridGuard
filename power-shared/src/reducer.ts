import { original, produce } from "immer";
import _ from "lodash";
import Prando from "prando";
import R from "ramda";
import { Action, ActionType } from "./interfaces/action";
import { Device, GridPowerConsumption, State, User } from "./interfaces/state";
import { unreachable } from "./util";

const historySize = 20;
export const tickMillis = 500;

export const nonSmartRatio = 0.8;

const gridYearlyEnergy = 58.1 * 1e9 * 1000; // Wh
export const gridMeanPower = gridYearlyEnergy / (365 * 24); // W

export function getPowerConsumption(
  tick: number,
  devices: Device[],
): GridPowerConsumption {
  let total = 0;
  let totalWithoutSavings = 0;
  const byDeviceClass: { [key: string]: number | undefined } = {};
  const byDeviceClassWithoutSavings: { [key: string]: number | undefined } = {};
  for (const deviceClassKey of allDeviceClassKeys) {
    byDeviceClass[deviceClassKey] = 0;
    byDeviceClassWithoutSavings[deviceClassKey] = 0;
  }

  const temp = gridMeanPower * getBasePowerConsumptionFromTick(tick);
  total += temp * nonSmartRatio;
  totalWithoutSavings += temp * nonSmartRatio;

  for (const device of devices) {
    total += device.powerConsumption;
    totalWithoutSavings += device.powerConsumptionWithoutSavings;
    byDeviceClass[device.deviceClassKey]! += device.powerConsumption;
    byDeviceClassWithoutSavings[device.deviceClassKey]! +=
      device.powerConsumptionWithoutSavings;
  }

  return {
    total,
    totalWithoutSavings,
    byDeviceClass,
    byDeviceClassWithoutSavings,
  };
}

function getBasePowerConsumptionFromTick(tick: number): number {
  return 1;
}

function getPowerProduction(
  tick: number,
  meanProduction: number,
  eventOngoing: boolean,
): number {
  return meanProduction * (eventOngoing ? 0.95 : 1);
}

function makeDeviceIndicesById(devices: Device[]): {
  [id: string]: number | undefined;
} {
  const deviceIndicesById: {
    [id: string]: number | undefined;
  } = {};
  for (const [i, device] of devices.entries()) {
    deviceIndicesById[device.id] = i;
  }
  return deviceIndicesById;
}

export const allDeviceClassKeys = new Set([
  "dishwasher",
  "fridge",
  "microwave",
  "oven",
  "light",
  "heater",
]);
export const allDeviceClassKeysSorted = R.sortBy(
  (v) => v,
  [...allDeviceClassKeys],
);

export function makeInitialState(): State {
  const meanProduction = gridMeanPower;
  const toggles: State["toggles"] = allDeviceClassKeysSorted.map((key) => ({
    key,
    targetSavingRatio: 0,
  }));

  const prando = new Prando();
  const devices: Device[] = [];
  for (let i = 0; i < 50; i++) {
    const deviceClassKey = prando.nextArrayItem(allDeviceClassKeysSorted);
    const id = prando.nextString(32);
    devices.push({
      id,
      deviceClassKey,
      powerConsumption: 3.14,
      powerConsumptionWithoutSavings: 3.14,
    });
  }

  const getPowerConsumptionHacked = (tick: number) => {
    const powerConsumption = getPowerConsumption(tick, devices);
    powerConsumption.total /= nonSmartRatio;
    return powerConsumption;
  };

  return {
    users: [],
    devices,
    deviceIndicesById: makeDeviceIndicesById(devices),
    meanProduction,
    toggles,
    autoAdjust: false,
    simulation: {
      tick: historySize,
      powerConsumption: getPowerConsumptionHacked(historySize),
      powerProduction: getPowerProduction(historySize, meanProduction, false),
    },
    simulationHistory: _.times(historySize, (i) => ({
      tick: i,
      powerConsumption: getPowerConsumptionHacked(i),
      powerProduction: getPowerProduction(i, meanProduction, false),
    })),
    eventOngoing: false,
  };
}

function makeInitialUser(id: string): User {
  return {
    id,
    seed: (new Prando(id) as any).hashCode(id), // HACK
  };
}

export const reducer = (_state: State, action: Action): State =>
  produce(_state, (state) => {
    switch (action.type) {
      case ActionType.AddUser: {
        if (state.users.find((u) => u.id === action.userId)) {
          throw new Error(`user already exists: ${action.userId}`);
        }
        state.users.push(makeInitialUser(action.userId));
        break;
      }
      case ActionType.RemoveUser: {
        state.users = state.users.filter((u) => u.id !== action.userId);
        break;
      }
      case ActionType.SetToggle: {
        for (const toggle of state.toggles) {
          if (toggle.key === action.key) {
            toggle.targetSavingRatio = action.targetSavingRatio;
            return;
          }
        }
        throw new Error(`toggle does not exist: ${action.key}`);
      }
      case ActionType.SetAutoAdjust: {
        state.autoAdjust = action.autoAdjust;
        break;
      }
      case ActionType.StartEvent: {
        state.eventOngoing = true;
        break;
      }
      case ActionType.EndEvent: {
        state.eventOngoing = false;
        break;
      }
      case ActionType.TickSimulation: {
        if (
          !action.devices.every((device) =>
            allDeviceClassKeys.has(device.deviceClassKey),
          )
        ) {
          throw new Error("invalid device class key");
        }
        state.devices = action.devices;
        state.deviceIndicesById = makeDeviceIndicesById(action.devices);
        state.meanProduction = _.mean(
          state.simulationHistory.map(
            (v) => v.powerConsumption.totalWithoutSavings,
          ),
        );
        state.simulationHistory.shift();
        state.simulationHistory.push(
          _.cloneDeep(original(state.simulation) ?? state.simulation),
        );
        state.simulation.tick++;
        state.simulation.powerConsumption = action.powerConsumption;
        state.simulation.powerProduction = getPowerProduction(
          historySize,
          state.meanProduction,
          state.eventOngoing,
        );
        if (state.autoAdjust) {
          const classKeysByPriority: string[][] = [
            ["light"],
            ["heater", "fridge"],
            ["microwave", "oven"],
            ["dishwasher"],
          ];
          const remainingClassKeys = new Set(allDeviceClassKeysSorted);
          for (const classKey of classKeysByPriority.flatMap((v) => v)) {
            remainingClassKeys.delete(classKey);
          }
          if (remainingClassKeys.size) {
            classKeysByPriority.push(
              R.sortBy((v) => v, [...remainingClassKeys]),
            );
          }

          const cumulativeConsumptionWithoutSavingsByPriority = [];
          let temp = 0;
          for (const classKeysThisPriority of classKeysByPriority) {
            const consumptionWithoutSavingsThisPriority = _.sum(
              classKeysThisPriority.map(
                (k) =>
                  state.simulation.powerConsumption.byDeviceClassWithoutSavings[
                    k
                  ],
              ),
            );
            temp += consumptionWithoutSavingsThisPriority;
            cumulativeConsumptionWithoutSavingsByPriority.push(temp);
          }

          const smartDeviceConsumptionWithoutSavings = _.sum(
            Object.values(
              state.simulation.powerConsumption.byDeviceClassWithoutSavings,
            ).map((v) => v ?? 0),
          );
          const nonSmartDeviceConsumption =
            state.simulation.powerConsumption.totalWithoutSavings -
            smartDeviceConsumptionWithoutSavings;

          let expectedConsumption = nonSmartDeviceConsumption;
          for (const classKeysThisPriority of classKeysByPriority) {
            const consumptionWithoutSavingsThisPriority = _.sum(
              classKeysThisPriority.map(
                (k) =>
                  state.simulation.powerConsumption.byDeviceClassWithoutSavings[
                    k
                  ],
              ),
            );
            const targetSavingAbsolute =
              expectedConsumption +
              consumptionWithoutSavingsThisPriority -
              state.simulation.powerProduction;
            let targetSavingRatio =
              targetSavingAbsolute / consumptionWithoutSavingsThisPriority;
            targetSavingRatio = Math.max(0, Math.min(1, targetSavingRatio));

            for (const classKey of classKeysThisPriority) {
              const toggle = state.toggles.find((t) => t.key === classKey);
              if (!toggle) {
                continue;
              }

              const oldTargetSavingRatio = toggle.targetSavingRatio;
              const maxStepSize = 0.03;
              toggle.targetSavingRatio = Math.max(
                oldTargetSavingRatio - maxStepSize,
                Math.min(oldTargetSavingRatio + maxStepSize, targetSavingRatio),
              );

              console.log("DEBUG", {
                classKey,
                toggleTargetSavingRatio: toggle.targetSavingRatio,
              });

              const expectedConsumptionThisClass =
                (1 - toggle.targetSavingRatio) *
                (state.simulation.powerConsumption.byDeviceClassWithoutSavings[
                  toggle.key
                ] ?? 0);
              expectedConsumption += expectedConsumptionThisClass;
            }
          }
        }
        break;
      }
      default: {
        unreachable(action);
      }
    }
  });
