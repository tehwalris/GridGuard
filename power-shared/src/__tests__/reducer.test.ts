import { Draft, produce } from "immer";
import R from "ramda";
import { Action, ActionType } from "../interfaces/action";
import { State } from "../interfaces/state";
import { makeInitialState, reducer } from "../reducer";

describe("reducer", () => {
  function makeReducerCustomCheckTest(
    modifyInputState: (state: State) => void,
    action: Action,
    check: (outputState: State, inputState: State) => void,
  ) {
    return () => {
      const inputState = makeInitialState();
      modifyInputState(inputState);
      const outputState = reducer(inputState, action);
      check(outputState, inputState);
    };
  }

  function makeReducerTest(
    modifyInputState: (state: State) => void,
    action: Action,
    modifyAsExpected: (state: Draft<State>) => void,
  ) {
    return makeReducerCustomCheckTest(
      modifyInputState,
      action,
      (outputState, inputState) => {
        const expectedOutputState = produce(inputState, (state) =>
          modifyAsExpected(state),
        );
        expect(outputState).toEqual(expectedOutputState);
      },
    );
  }

  function makeReducerErrorTest(
    modifyInputState: (state: State) => void,
    action: Action,
    expectedError: RegExp,
  ) {
    return () => {
      const inputState = makeInitialState();
      modifyInputState(inputState);
      expect(() => reducer(inputState, action)).toThrowError(expectedError);
    };
  }

  describe("AddUser", () => {
    test(
      "add user",
      makeReducerTest(
        () => {},
        {
          type: ActionType.AddUser,
          userId: "someId",
        },
        (state) => {
          state.users.push({
            id: "someId",
            seed: -1869716294,
          });
        },
      ),
    );
  });

  describe("SetToggle", () => {
    test(
      "set existing toggle",
      makeReducerTest(
        (state) => {
          state.toggles[2].powered = false;
        },
        {
          type: ActionType.SetToggle,
          key: "microwave",
          powered: true,
        },
        (state) => {
          state.toggles[2].powered = true;
        },
      ),
    );

    test(
      "setting missing toggle fails",
      makeReducerErrorTest(
        (state) => {
          state.toggles[2].powered = false;
        },
        {
          type: ActionType.SetToggle,
          key: "laser-cannon",
          powered: true,
        },
        /toggle does not exist/,
      ),
    );
  });

  describe("TickSimulation", () => {
    test(
      "tick simulation",
      makeReducerCustomCheckTest(
        () => {},
        {
          type: ActionType.TickSimulation,
        },
        (outputState, inputState) => {
          expect(outputState.simulation.tick).toBe(
            inputState.simulation.tick + 1,
          );
          expect(outputState.simulation.powerConsumption).not.toBe(
            inputState.simulation.powerConsumption,
          );
          expect(outputState.simulationHistory.length).toBe(
            inputState.simulationHistory.length,
          );
          expect(outputState.simulationHistory.slice(0, -1)).toStrictEqual(
            inputState.simulationHistory.slice(1),
          );
          expect(R.last(outputState.simulationHistory)!).toStrictEqual(
            inputState.simulation,
          );
        },
      ),
    );
  });
});
