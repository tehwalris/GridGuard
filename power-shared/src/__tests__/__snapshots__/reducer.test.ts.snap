// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`reducer StartGameAction 1`] = `
{
  "distributor": 2,
  "phase": "Distributing",
  "pieceStartTime": 1614527659000,
  "toDistribute": [
    [
      34,
      90,
      66,
      70,
      52,
      39,
      55,
    ],
    [
      78,
      87,
      76,
      28,
      72,
      53,
      2,
    ],
    [
      30,
      15,
      58,
      98,
      96,
      45,
      83,
    ],
    [
      79,
      54,
      94,
      77,
      26,
      32,
      37,
    ],
  ],
  "totalToDistribute": 28,
}
`;

exports[`reducer StartGameAction 2`] = `
[
  [
    [
      48,
      47,
      16,
      20,
      4,
      19,
      6,
      5,
      46,
      18,
    ],
    [
      43,
      33,
      17,
      42,
      3,
      1,
      14,
      0,
      44,
      29,
    ],
    [
      31,
      null,
      null,
      null,
      null,
      null,
      null,
      null,
      null,
      null,
    ],
    [
      null,
      null,
      null,
      null,
      null,
      null,
      null,
      null,
      null,
      null,
    ],
    [
      null,
      null,
      null,
      null,
      null,
      null,
      null,
      null,
      null,
      null,
    ],
  ],
  [
    [
      49,
      35,
      38,
      8,
      12,
      24,
      22,
      25,
      21,
      41,
    ],
    [
      51,
      9,
      27,
      23,
      50,
      40,
      10,
      36,
      7,
      13,
    ],
    [
      11,
      null,
      null,
      null,
      null,
      null,
      null,
      null,
      null,
      null,
    ],
    [
      null,
      null,
      null,
      null,
      null,
      null,
      null,
      null,
      null,
      null,
    ],
    [
      null,
      null,
      null,
      null,
      null,
      null,
      null,
      null,
      null,
      null,
    ],
  ],
  [
    [
      56,
      103,
      99,
      89,
      61,
      88,
      104,
      86,
      62,
      84,
    ],
    [
      71,
      75,
      74,
      102,
      101,
      59,
      57,
      60,
      73,
      85,
    ],
    [
      100,
      null,
      null,
      null,
      null,
      null,
      null,
      null,
      null,
      null,
    ],
    [
      null,
      null,
      null,
      null,
      null,
      null,
      null,
      null,
      null,
      null,
    ],
    [
      null,
      null,
      null,
      null,
      null,
      null,
      null,
      null,
      null,
      null,
    ],
  ],
  [
    [
      65,
      68,
      81,
      67,
      105,
      111,
      91,
      93,
      82,
      106,
    ],
    [
      97,
      109,
      63,
      108,
      107,
      80,
      95,
      92,
      64,
      110,
    ],
    [
      69,
      null,
      null,
      null,
      null,
      null,
      null,
      null,
      null,
      null,
    ],
    [
      null,
      null,
      null,
      null,
      null,
      null,
      null,
      null,
      null,
      null,
    ],
    [
      null,
      null,
      null,
      null,
      null,
      null,
      null,
      null,
      null,
      null,
    ],
  ],
]
`;
